docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,453 @@
+  because we only want to compute gradients with respect to `params` and not the
+  entire `variables`.
+- We can still log internal optimizer state, such as the learning rate, as long
+  as Optax transformations expose that information in their respective state.
+  For example, `optax.scale_by_schedule()` currently only exposes
+  `opt_state.count` but could easily be extend to also expose the `step_size`.
+  The same is true for internal optimizer states that change over time.
+
+## Multi Optimizer
+[Multi Optimizer]: #multi-optimizer
+
+The [previous API] defined `flax.optim.MultiOptimizer` for processing different
+parts of the parameter tree with different optimizers:
+
+```python
+biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: path.endswith('/bias'))
+not_biases_traversal = flax.optim.ModelParamTraversal(
+    lambda path, _: not path.endswith('/bias'))
+
+optimizer_def = flax.optim.MultiOptimizer(
+    (biases_traversal, flax.optim.GradientDescent(learning_rate=0.1)),
+    (not_biases_traversal, flax.optim.GradientDescent(learning_rate=0.05)),
+)
+```
+
+Note how we first define a traversal that selects parameters based on their
+path (which is the concatenation of module scopes and variable name), and then
+create a `MultiOptimizer` that binds a different optimizer for each of these
+separate traversals.
+
+Optax has recently implemented `optax.masked()` that can be used for specifying
+gradient transformations that only applied to a subset of the gradients:
+
+```python
+def dict_map_with_path(fn, d, path=''):  # a crude "ModelParamTraversal"
+    if isinstance(d, dict):
+        return {k: dict_map_with_path(fn, v, path+'/'+k) for k, v in d.items()}
+    return fn(path, d)
+
+is_bias_mask = dict_map_with_path(
+    lambda path, value: path.endswith('/bias'), variables['params'])
+not_bias_mask = jax.tree_map(lambda x: not x, is_bias_mask)
+
+tx = optax.chain(
+    optax.masked(optax.sgd(learning_rate=0.1), mask=is_bias_mask),
+    optax.masked(optax.sgd(learning_rate=0.05), mask=not_bias_mask),
+)
+```
+
+Remarks:
+
+- Above `dict_map_with_path()` emulates `flax.optim.ModelParamTraversal`.
+- Note that now we need `variables['params']` already when the gradient
+  transformation is defined, i.e. before calling `tx.init()`.
+- Note the change in behavior when a parameter matches multiple masks: With
+  `flax.optim.ModelParamTraversal` all but the last match would be discarded.
+  With `optax.chain(optax.masked())` the gradient updates are simply chained
+  and if a gradient matches multiple masks then all those transformations are
+  applied.
+
+## Train State
+[Train State]: #train-state
+
+In Flax it is common to hand around a `TrainState` object that can then be
+used for checkpointing. This simplifies the above [Optax training step] a bit by
+reducing the number of arguments and getting rid of the `static_argnums`.
+
+```python
+class TrainState(flax.struct.PyTreeNode):
```

Comment:
I think composition is better than inheritance here.
This is very similar to the old idea of having an optimizer dataclass:
```
class Optimizer(PyTreeNode):
  params: flax.core.FrozenDict[str, Any]
  model_state: flax.core.FrozenDict[str, Any]
  tx: optax.GradientTransformation = flax.struct.field(pytree_node=False)
  opt_state: optax.OptState

  def apply_gradient(self, grads):  # should apply_gradient take a model_state update?
    updates, new_opt_state = self.tx.update(
      grads, self.opt_state, self.params)
    new_params = optax.apply_updates(self.params, updates)
    return self.replace(params=new_params, opt_state=new_opt_state)

  @classmethod
  def create(cls, variables, tx):
     ...
```

inheritance implies that the TrainState only has a single model & optimizer. Also dataclasses don't behave very well under inheritance in general.

Another question is here whether you should store variables or params and model_state separately. I don't have a strong opinion here. We could even define the other one using `@property`. To me it seems splitting is a little bit easier because you only ever need to merge again inside the loss function and it's easy to accidentally use `variables` instead of `variables.copy(params=params_inside_loss_fn)`.

Reply:
+1 for storing params and model_state separately.

Reply:
Yeah I think composition is much better here. Updated.

I decided to include `model_state` in the signature to make it harder to forget that. Since the signature is different I thought it's also better to use a new name (going for `update()` as in `GradientTransformation.update()`).